lib/Cron/Workers/SendingQueue/SendingThrottlingHandler.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+namespace MailPoet\Cron\Workers\SendingQueue;
+
+use MailPoet\Logging\LoggerFactory;
+use MailPoet\Settings\SettingsController;
+use MailPoet\WP\Functions as WPFunctions;
+use MailPoetVendor\Monolog\Logger;
+
+class SendingThrottlingHandler {
+  public const BATCH_SIZE = 20;
+  public const SETTINGS_KEY = 'mta_throttling';
+  public const SUCCESS_THRESHOLD_TO_INCREASE = 10;
+
+  /** @var Logger */
+  private $logger;
+
+  /** @var SettingsController */
+  private $settings;
+
+  /** @var WPFunctions */
+  private $wp;
+
+  public function __construct(
+    LoggerFactory $loggerFactory,
+    SettingsController $settings,
+    WPFunctions $wp
+  ) {
+    $this->logger = $loggerFactory->getLogger(LoggerFactory::TOPIC_MSS);
+    $this->settings = $settings;
+    $this->wp = $wp;
+  }
+
+  public function getBatchSize(): int {
+    $throttlingSettings = $this->loadSettings();
+    if (isset($throttlingSettings['batch_size'])) {
+      return $throttlingSettings['batch_size'];
+    }
+    return $this->getMaxBatchSize();
+  }
+
+  private function getMaxBatchSize(): int {
+    return $this->wp->applyFilters('mailpoet_cron_worker_sending_queue_batch_size', self::BATCH_SIZE);
+  }
+
+  public function throttleBatchSize(): int {
+    $batchSize = $this->getBatchSize();
+    if ($batchSize > 1) {
+      $batchSize = (int)ceil($this->getBatchSize() / 2);
+      $throttlingSettings['batch_size'] = $batchSize;
+      $this->logger->error("MailPoet throttling: decrease batch_size to: {$batchSize}");
+      $this->saveSettings($throttlingSettings);
+    }
+
+    return $batchSize;
+  }
+
+  public function processSuccess(): void {
+    $throttlingSettings = $this->loadSettings();
+    if (!isset($throttlingSettings['batch_size'])) {
+      return;
+    }
+    $throttlingSettings['success_in_row'] = isset($throttlingSettings['success_in_row']) ? ++$throttlingSettings['success_in_row'] : 1;
+    $this->logger->info("MailPoet throttling: increase success_in_row to: {$throttlingSettings['success_in_row']}");
+    if ($throttlingSettings['success_in_row'] === self::SUCCESS_THRESHOLD_TO_INCREASE) {
```

Comment:
I would rather use `>=`. Right now === seems to be safe but we may introduce some changes in the future combined with the fact cron may run by accident in parallel and if we for some reason get over SUCCESS_THRESHOLD_TO_INCREASE we would never return to a bigger batch.